python/rikai/parquet/dataset.py

```diff
@@ -140,13 +151,35 @@ def _convert(
         return converted
 
     def __iter__(self):
+        shuffler = RandomShuffler(
+            self.shuffler_capacity if self.shuffle else 1, self.seed
+        )
+        group_count = 0
         for filepath in self.files:
             fs, path = FileSystem.from_uri(filepath)
             with fs.open_input_file(path) as fobj:
                 parquet = pg.ParquetFile(fobj)
                 for group_idx in range(parquet.num_row_groups):
+                    # A simple form of row-group level bucketing without memory overhead.
+                    # Pros:
+                    #  - It requires zero communication to initialize the distributed policy
+                    #  - It uses little memory and no startup overhead, i.e. collecting row groups.
+                    # Cons:
+                    #   The drawback would be if the world size is much larger than
+                    #   the average number of row groups. As a result, many of the
+                    #   file open operations would be wasted.
+                    group_count += 1
+                    if group_count % self.world_size != self.rank:
+                        continue
                     row_group = parquet.read_row_group(group_idx, columns=self.columns)
                     for batch in row_group.to_batches():  # type: RecordBatch
                         # TODO: read batches not using pandas
                         for _, row in batch.to_pandas().iterrows():
-                            yield self._convert(row.to_dict(), self.spark_row_metadata)
+                            shuffler.append(row)
+                            # Control the capacity inexplicitily.
```

Comment:
implicitly? what do you mean here?

Reply:
the shuffler itself does not block `append` once the shuffler is full.  This line will maintain the capacity to approximately full (`capacity +/- 1/0`) by the caller instead of the shuffler itself. 

i will reword this.